core_lib/src/interface/editor.cpp

```diff
@@ -508,108 +502,193 @@ void Editor::resetAutoSaveCounter()
-        if (vectorImage == nullptr) { return; }
-        g_clipboardVectorImage = *vectorImage;  // copy the image
+        currentLayer->deselectAll();
     }
 }
 
 void Editor::paste()
 {
-    Layer* layer = mObject->getLayer(layers()->currentLayerIndex());
-    if (layer != nullptr)
-    {
-        if (layer->type() == Layer::BITMAP && g_clipboardBitmapImage.image() != nullptr)
-        {
-            backup(tr("Paste"));
+    Layer* currentLayer = layers()->currentLayer();
 
-            BitmapImage tobePasted = g_clipboardBitmapImage.copy();
-            qDebug() << "to be pasted --->" << tobePasted.image()->size();
-            if (select()->somethingSelected())
-            {
-                QRectF selection = select()->mySelectionRect();
-                if (g_clipboardBitmapImage.width() <= selection.width() && g_clipboardBitmapImage.height() <= selection.height())
-                {
-                    tobePasted.moveTopLeft(selection.topLeft());
-                }
-                else
-                {
-                    tobePasted.transform(selection, true);
-                }
-            }
-            auto pLayerBitmap = static_cast<LayerBitmap*>(layer);
-            mScribbleArea->handleDrawingOnEmptyFrame();
-            BitmapImage *bitmapImage = pLayerBitmap->getLastBitmapImageAtFrame(currentFrame(), 0);
-            Q_CHECK_PTR(bitmapImage);
-            bitmapImage->paste(&tobePasted); // paste the clipboard
-        }
-        else if (layer->type() == Layer::VECTOR && clipboardVectorOk)
-        {
+    Q_ASSERT(currentLayer != nullptr);
+
+    if (clipboards()->getClipboardFrames().empty()) {
+
+        if (clipboards()->canPaste(currentLayer)) {
             backup(tr("Paste"));
-            deselectAll();
-            mScribbleArea->handleDrawingOnEmptyFrame();
-            VectorImage* vectorImage = static_cast<LayerVector*>(layer)->getLastVectorImageAtFrame(currentFrame(), 0);
-            Q_CHECK_PTR(vectorImage);
-            vectorImage->paste(g_clipboardVectorImage);  // paste the clipboard
-            select()->setSelection(vectorImage->getSelectionRect(), false);
         }
+
+        clipboards()->updateIfNeeded(currentLayer);
+
+        BitmapImage clipboardImage = clipboards()->getBitmapClipboard();
+        VectorImage clipboardVectorImage = clipboards()->getVectorClipboard();
+        if (currentLayer->type() == Layer::BITMAP && clipboardImage.isLoaded()) {
+            pasteToCanvas(&clipboardImage, mFrame);
+        } else if (currentLayer->type() == Layer::VECTOR && clipboardVectorImage.isValid()) {
+            pasteToCanvas(&clipboardVectorImage, mFrame);
+        }
+    } else {
+        pasteToFrames();
     }
-    mScribbleArea->updateCurrentFrame();
+
+    emit layers()->currentLayerChanged(layers()->currentLayerIndex());
+    mScribbleArea->updateFrame(mFrame);
+}
+
+bool Editor::canCopy() const
```

Comment:
I personally think it makes sense because you're dependant on everything from editor anyway, so why not simply make it a method with no arguments and make it easier to use instead of of mEditor->canCopy(mEditor->getCurrentFrame, mEditor->layers()->currentLayer)?

Reply:
Hmm so you mean that it's given that we always copy and paste from the current frame and layer? I could accept that argument, then I'll remove the parameters.

Reply:
> Hmm so you mean that it's given that we always copy and paste from the current frame and layer? I could accept that argument, then I'll remove the parameters.

@CandyFace In the sense of working from and to an active layer, pasting a sequence of frames will occur from the initial frame we set the playhead at, however shouldn't copying / cutting be allowed to have the playhead positioned at any frame? I mean what should matter is the selection itself not the current frame position in that case 🤔 

The simplest example use case I can think of is I could select a non-juxtaposed sequence of frames to cut (i.e leave the even frame drawings, select the odd frames), and then scrub through them to see If I missed any so the playhead would be at the middle or the end of the selection. Sometimes, for efficiency, I could also move the playhead to the place I want to paste, but since the frames are selected I press CTRL + C assuming the selection takes priority over the playhead position.

What do you think?

Reply:
When you move the playhead position, you evidentially also change the current frame, so pasting will always occurs on the frame you stand on, not where you copied. The position of the playhead has no role in this, except when you paste the frames, which happens on the current frame.

Copying and pasting across multiple layers is still not a thing, and will not become a thing in this PR, so I think it's fine to say that the copy/paste logic assumes that we always make changes to the current layer.